sdk/src/utils/pdf_utils.rs

```diff
@@ -0,0 +1,403 @@
+                EMBEDDED_FILES_KEY => Reference(embedded_files_ref)
+            });
+
+            // Set /Names key in `Catalog` to reference above indirect object names dictionary.
+            self.document.catalog_mut()?.set(NAMES_KEY, names_ref);
+        }
+
+        Ok(())
+    }
+
+    /// Writes the provided bytes to the PDF as a `FileAttachment` `Annotation`. This `Annotation`
+    /// is added to the first page of the `PDF`, to the lower left corner.
+    fn write_manifest_as_annotation(&mut self, bytes: Vec<u8>) -> Result<(), Error> {
+        let file_stream_reference = self.add_c2pa_embedded_file_stream(bytes);
+        let file_spec_reference = self.add_embedded_file_specification(file_stream_reference);
+
+        self.set_af_relationship(file_spec_reference)?;
+        self.add_file_attachment_annotation(file_spec_reference)?;
+
+        Ok(())
+    }
+}
+
+impl Pdf {
+    /// Adds the C2PA `Annotation` to the PDF.
+    ///
+    /// ### Note:
+    /// The `FileAttachment` annotation is added to the first page of the PDF in the lower
+    /// left-hand corner. The `FileAttachment`'s location is not defined in the spec as of version
+    /// `1.3`.
+    fn add_file_attachment_annotation(
+        &mut self,
+        file_spec_reference: ObjectId,
+    ) -> Result<(), Error> {
+        let annotation = dictionary! {
+            "Type" => Name("Annot".into()),
+            "Contents" => Object::string_literal(CONTENT_CREDS),
+            "Name" => Object::string_literal(CONTENT_CREDS),
+            SUBTYPE_KEY => Name("FileAttachment".into()),
+            "FS" => Reference(file_spec_reference),
+            // Places annotation in the lower left-hand corner. The icon will be 10x10.
+            "Rect" => vec![0.into(), 0.into(), 10.into(), 10.into()],
+        };
+
+        // Insert annotation into the PDF.
+        let annotation_ref = self.document.add_object(annotation);
+
+        // Find the first page of the PDF.
+        let first_page_ref = self
+            .document
+            .page_iter()
+            .next()
+            .ok_or_else(|| Error::AddingAnnotation)?;
+
+        // Get the annotations of the first page of the PDF and add the reference to the newly
+        // inserted annotation.
+        let mut annotations = self.get_page_annotations(first_page_ref)?;
+        annotations.push(Reference(annotation_ref));
+
+        // Update the first page's `Annots` value with the updated annotations array.
+        let first_page_dict = self.document.get_dictionary_mut(first_page_ref)?;
+        first_page_dict.set(ANNOTATIONS_KEY, annotations.to_vec());
+
+        Ok(())
+    }
+
+    /// Gets the `Annotation`s for the provided Page `ObjectId`.
+    ///
+    /// This function will create an empty vector of annotations if the associated page of the PDF
+    /// doesn't have a vector of annotations already defined.
+    fn get_page_annotations(&self, page_ref: ObjectId) -> Result<Vec<Object>, Error> {
```

Comment:
Working on a way to get rid of the `clones` below, but I'm running into the same issue above where the borrow checker complains of multiple mutable references to the `lopdf::Document`.

Reply:
Do you need to return owned `Vec<Object>` here? Is there any reason the success side of this `Result` couldn't be `Option<[&Object]>` (with some lifetime as needed)?

Reply:
I added this to make unit testing easier in manifest generation code. It's hard to make PDFs that cause some functions in this trait fail. For example, `Pdf::write_manifest_as_annotation` calls `lopdf` code that can fail, but it's hard to generate a PDF that cause that to fail. Using a trait, we can instead use a mocking tool like `mockall` to do something like this: 

```rust
let mut pdf = MockC2paPdf::default();
pdf
    .expect_write_manifest_as_annotation()
    .returning(|_| Err(lopdf::Error::Blah))

// etc...
```

Happy to remove if you think it's not necessary though!